tests/dragonfly/cluster_test.py

```diff
@@ -907,272 +972,196 @@ async def test_random_keys():
+    assert await nodes[1].client.get("KEY13") == "value"
+    assert await nodes[1].client.get("KEY14") == "value"
+    assert await nodes[1].client.get("KEY15") == "value"
+    assert await nodes[1].client.get("KEY16") == "value"
+    assert await nodes[1].client.get("KEY17") == "value"
+    assert await nodes[1].client.get("KEY18") == "value"
+    assert await nodes[1].client.get("KEY19") == "value"
+    assert await nodes[1].client.get("KEY20") == "value"
+    assert await nodes[1].client.get("KEY21") == "value"
+    assert await nodes[1].client.execute_command("DBSIZE") == 19
+
+    await check_for_no_state_status([node.admin_client for node in nodes])
+    await close_clients(*[node.client for node in nodes], *[node.admin_client for node in nodes])
 
-    assert (
-        await c_nodes_admin[1].execute_command("DFLYCLUSTER", "SLOT-MIGRATION-STATUS") == "NO_STATE"
-    )
-    assert (
-        await c_nodes_admin[0].execute_command("DFLYCLUSTER", "SLOT-MIGRATION-STATUS") == "NO_STATE"
-    )
 
-    await close_clients(*c_nodes, *c_nodes_admin)
+@dfly_args({"proactor_threads": 4, "cluster_mode": "yes"})
+async def test_network_disconnect_during_migration(df_local_factory, df_seeder_factory):
+    instances = [
+        df_local_factory.create(port=BASE_PORT + i, admin_port=BASE_PORT + i + 1000)
+        for i in range(2)
+    ]
 
+    df_local_factory.start_all(instances)
 
-@dataclass
-class MigrationInfo:
-    ip: str
-    port: int
-    slots: list
-    node_id: str
+    nodes = [(await create_node_info(instance)) for instance in instances]
+    nodes[0].slots = [(0, 16383)]
+    nodes[1].slots = []
 
+    await push_config(json.dumps(generate_config(nodes)), [node.admin_client for node in nodes])
 
-@dataclass
-class NodeInfo:
-    instance: DflyInstance
-    client: aioredis.Redis
-    admin_client: aioredis.Redis
-    slots: list
-    next_slots: list
-    migrations: list
-    id: str
+    seeder = df_seeder_factory.create(keys=30000, port=nodes[0].instance.port, cluster_mode=True)
 
+    await seeder.run(target_deviation=0.1)
 
-async def create_node_info(instance):
-    admin_client = instance.admin_client()
-    ninfo = NodeInfo(
-        instance=instance,
-        client=instance.client(),
-        admin_client=admin_client,
-        slots=[],
-        next_slots=[],
-        migrations=[],
-        id=await get_node_id(admin_client),
-    )
-    return ninfo
+    proxy = Proxy("127.0.0.1", 1111, "127.0.0.1", nodes[1].instance.port)
+    await proxy.start()
+    task = asyncio.create_task(proxy.serve())
 
+    nodes[0].migrations.append(MigrationInfo("127.0.0.1", proxy.port, [(0, 16383)], nodes[1].id))
+    try:
+        logging.debug("Start migration")
+        await push_config(json.dumps(generate_config(nodes)), [node.admin_client for node in nodes])
+
+        for _ in range(10):
+            await asyncio.sleep(random.randint(0, 10) / 20)
+            logging.debug("drop connections")
+            proxy.drop_connection()
```

Comment:
what happens here? we drop one connection, does all migration starts from scratch?
can you check that slot migration status is in reconnecting?
is the source node creating new connection and this is the reason that the migration is continuing after drop connection?

Reply:
we drop all connections. we can not exactly say what the current status of the connections and some can restart from some point (like main connection that does INIT and ACK), but FLOWs connections need to restart from scratch.
The source node creates new connections. Migration continues only if the main connections are dropped but if FLOWs connections are dropped we restart the whole process. We can not check for reconnecting status because the status can be any of the CONNECTING, SYNC, FINISHED.

Reply:
unfortunately, I don't know, so I run drop connections 10 times with different sleep.

Reply:
can you do a check that the migration status is not in no-state or finished before calling drop connection?

Reply:
there is no sense because it can finish after the check and it doesn't provide some guarantees